### ISSUE TITLE
Cache gets corrupted when running `pod install` concurrently

### ISSUE THREAD

**Issue by codiophile** [CocoaPods/CocoaPods#10013]
* [x] I've read and understood the [*CONTRIBUTING* guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md). 🌈

# Report

## What did you do?

Ran `pod install` concurrently from different jobs on the build server.

(We don't know for sure that concurrency is the cause, but it's intermittency and the fact that multiple processes are writing to the same location at the same time, suggests that it probably is.)

## What did you expect to happen?

That the cache and pod installations would be fine.

## What happened instead?

Sometimes the cache gets corrupted and caches for some pods were emptied and pod install now installs those pods as empty directories.

## CocoaPods Environment

(Not allowed to share the whole thing as it reveals details of our project. Below is what looks relevant and safe to share.)

### Stack

```
   CocoaPods : 1.8.3
        Ruby : ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]
    RubyGems : 3.1.2
        Host : Mac OS X 10.15.6 (19G73)
       Xcode : 11.1 (11A1027)
         Git : git version 2.18.0
Ruby lib dir : /opt/jenkins_tools/ruby/2.7.1/lib
```

### Installation Source

```
Executable Path: /Users/jenkins/.bundle/ruby/2.7.0/bin/pod
```

### Plugins

```
cocoapods-deintegrate : 1.0.4
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-stats       : 1.1.0
cocoapods-trunk       : 1.4.1
cocoapods-try         : 1.1.0
```

## Additional Background

This is a long standing issue in our environment. We run multiple jobs in parallel per build server to better utilise our capacity. Most of the time, everything is fine and works as expected, but every now and then the jobs will fail because of missing files in the `Pods/` directory. Once a build server has a corrupted cache, the builds will fail until we delete the cache, so that the pods are re-downloaded and copied correctly to the `Pods/` directory.

We don't know for sure, but our theory is that this is because we are running multiple instances of `pod install` on the same machine using the same cache and that under certain circumstances that corrupts the cache. Until now, since it is relatively infrequent, we've just been manually clearing the caches when they got corrupted, but we recently saw an increase in frequency of cache corruptions and decided to try to fix it.

Work is already under way to refactor our build scripts and pipelines to eliminate this concurrency and we're hoping that will fix the cache corruptions. Because there are multiple apps and multiple teams, someone could inadvertently add a `pod install` in their build script that risks running concurrently and break the cache anyway. So we can really only decrease the probability of future cache corruptions.

The best solution would be if CocoaPods had built in support for running `pod install` concurrently. It should take care of locking the cache, so that multiple instances don't try to write to the cache at the same time. I suppose we don't know for sure that concurrency is the issue, but it seems like the most likely cause. If there's some other error condition that corrupts the cache, we'd obviously like that to be fixed.

We'd be very interested in hearing about whether others have experienced the same problem with caches getting corrupted.

**Comment by dnkoutso** [comment]
Sounds like this would happen yes most likely. I think if possible a deeper dive into where exactly this can happen will help us pin point where to actually fix it.

**Comment by amorde** [comment]
I think this is a dupe of #6280

There was a PR for that which was closed due to inactivity. If anyone wants to revive that we'd be happy to review it 👍

**Comment by codiophile** [comment]
Thanks @amorde. I didn't find that one in my search. I'm assuming you're talking about #6535. Was there an issue with that pull request or was it just that the author became unresponsive? I wouldn't mind trying to revive it. I suppose it's so outdated now that I'm better off creating a new pull request.

**Comment by amorde** [comment]
It would be great if tests were included as well. I think the general approach is good with the caveat that there may be other places in CocoaPods that don't handle concurrent invocations, so there may be more work to do to support this. Incremental progress is still progress!

**Comment by codiophile** [comment]
@armode Trying to follow the instructions for contributing:
https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md
https://guides.cocoapods.org/contributing/contribute-to-cocoapods.html

But I'm getting a lot of errors. The [Rainforest repo](https://github.com/CocoaPods/Rainforest) hasn't been touched in over a year and the build is failing. I'm getting the same error locally, so it's clearly not a problem with my local setup. The instructions for how to contribute, seem to be even more outdated than the [Rainforest repo](https://github.com/CocoaPods/Rainforest).

Even when checking out and running the tests for this repo in isolation, I'm getting test failures. I suppose this could be due to some configuration missing locally, since it is passing on Travis CI. Actually, looking at the failures, it seems like the problem is that my version of git isn't giving the same output as you are expecting. What version of git is required for the tests to pass? I suppose the best thing to do is to change those tests to use a regex that passes for all recent versions of git, but that's not the issue I wanted to fix. :)

Do you have any general guidance on how to proceed? Should I just disregard the [Rainforest repo](https://github.com/CocoaPods/Rainforest) and just work on this repository directly? Considering that the [Rainforest repo](https://github.com/CocoaPods/Rainforest) is broken, I'm assuming that you're not using it anymore? How is your local environment set up?

**Comment by codiophile** [comment]
Despite the issues with my local environment, I created a pull request: #10018

I went a bit further than the old pull request, by covering both reads and writes. Otherwise there is a risk that one process is copying from the cache while another process is deleting the cache. I also made it a bit easier to use by encapsulating the whole locking logic in user friendly methods.

I didn't add any tests to verify that concurrent installations don't mess up the cache, but the code that I've touched is already covered by tests and they are all passing, so at the very least, I didn't break anything and at least in theory concurrent installations shouldn't corrupt the cache anymore.

I noticed that `ensure_matching_version` deletes the whole cache if it was created by a different version of CocoaPods. So when building different projects/branches concurrently, there could still be an issue if they aren't all using the same version of CocoaPods. I wanted to fix that as well, but I'm not sure what the correct behaviour is. I can create a separate issue to discuss that. It's anyway an edge case, so it's not really that urgent. Most of the time, we're not using different versions of CocoaPods, and we do have the power to ask project maintainers to use the same specific version of CocoaPods. We could also have different cache directories for different projects if necessary.

Please let me know if my change can be merged and how soon it can be released. :)

**Comment by dnkoutso** [comment]
thank you! we can review and land this for 1.11.0.

**Comment by amorde** [comment]
@codiophile apologies for the dev setup issues - that guide is a bit old and doesn't reflect how we develop today.

I'd recommend this flow:

```
git clone git@github.com:CocoaPods/CocoaPods.git
cd CocoaPods
which bundle || sudo gem install bundler
bundle install
bundle exec rake spec:unit # unit tests only
bundle exec rake spec:integration # integration tests only
bundle exec rake spec # all tests
bundle exec pod install --project-directory=/some/path/to/test/project # test current code against a project
```

> Actually, looking at the failures, it seems like the problem is that my version of git isn't giving the same output as you are expecting.

Yes this is a known issue, CI has an older version of git and we haven't updated CI to use a newer one. The current workaround is to ignore those changes & failures.

> Considering that the Rainforest repo is broken, I'm assuming that you're not using it anymore? How is your local environment set up?

We use Rainforest for releasing new versions of CocoaPods & related plugins, but not for day-to-day development
